### DISCUSSION TITLE
[Android] Seminar 1 ppt에 대한 질문

### DISCUSSION THREAD

**Discussion by kkomul1** [wafflestudio/seminar-2023#163] [Q&A] [answered]
PPT 관련 질문입니다. 아직 끝까지 다 보지 못해서 질문이 더 있을 수 있습니다. 너무 질문이 많은 것 같아 죄송합니다..ㅠㅠ

1. p.10에서 lateinit을 통해 binding을 나중에 초기화하겠다고 명시한 후 onCreate()가 실행될 때 초기화를 하고 있는데, 굳이 그렇게 해야하는 이유를 모르겠습니다. Activity가 onCreate()되든 안되든 xml 파일은 이미 존재하는 것인데, 그러면 ActivityMainBinding class도 생성되어 있어서 binding을 바로 초기화할 수 있는 것 아닌가요? 굳이 나중에 초기화하는 이유가 궁금합니다.
2. p.16의 3에서, adapter를 만들 때 ArrayAdapter라는 class를 이용해서 만든 것 같은데, 저 class가 무엇을 하는 것인지, adapter로 지정할 수 있는 class에는 어떤 것들이 있는지 궁금합니다.
3. p.16의 4에서처럼, 각 view에는 adapter라는게 있는 것 같습니다. (아니면 View를 extend하는 ListView class에만 adapter라는 field가 있는건가요?) 각 view에 있는 adapter가 하는 역할이 무엇인가요? 데이터와 UI를 매핑시켜 준다는 말이 잘 이해가 안되는 것 같습니다.
4. p.17에서, 생성자의 매개변수의 type이 ArrayList<T>라고 되어있는데, ArrayList와 같은 type들은 저희가 다 지정해 주어야 하나요? 아니면 그냥 kotlin의 type inference에 맡기면 될까요? ppt 예시와 같은 경우에는 ArrayList<T> 대신 List<T>를 사용해야 되지 않을까요?
5. p.17에서 영상을 보면 private val context: Context를 생성자의 매개변수로 추가하시던데, 그렇게 하는 이유가 무엇인가요? context를 MyListAdapter class에서 사용하지도 않으셔서 왜 굳이 생성자에서 받으신건지 궁금합니다.
6. p.17의 getView 함수에서, position, convertView, parent는 함수의 매개변수로 받으면서 왜 사용하지 않는건가요? 그리고 position은 list의 index 정도로 이해했는데, convertView와 parent가 의미하는 바가 무엇인가요? 마지막으로 parent의 type인 ViewGroup class가 무엇인가요?

추가질문: Discussions에 질문을 남길 때 Labels와 Milestones는 어떻게 추가하나요..? Issues에 있는 것과 UI가 달라서 질문드립니다.
![image](https://github.com/wafflestudio/seminar-2023/assets/111198830/4e4ce4cd-3b05-47e1-a7de-1c088b8a2876)

**Comment by JuTaK97** [comment, marked as answer]
## 1. view binding을 late로 초기화하는 이유
말씀하신 대로 ActivityMainBinding는 이미 만들어져 있습니다. 하지만 inflate할 때 필요한 `layoutInflator` 가 문제입니다.
실제로 
```kotlin
class MainActivity : AppCompatActivity() {

    private val binding: ActivityMainBinding = ActivityMainBinding.inflate(layoutInflater)
    private val viewModel: MainViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)
```
이렇게 해 보면 이런 에러가 납니다.
<img width="995" alt="image" src="https://github.com/wafflestudio/seminar-2023/assets/88367636/165e937a-66e0-4df6-a9b7-da2163b0a8b6">

이유가 뭘까요? layoutInflator를 파고 들어가 보면 Window.java의 이 코드에 도달합니다.
```
    /**
     * Quick access to the {@link LayoutInflater} instance that this Window
     * retrieved from its Context.
     *
     * @return LayoutInflater The shared LayoutInflater.
     */
    @NonNull
    public abstract LayoutInflater getLayoutInflater();
```
LayoutInflator를 가져오려면 올바른 context가 있어야 하는데, MainActivity 클래스 생성 시점에는 context가 null입니다. context는 프레임워크가 onCreate를 불러 줄 시점에서야 가져올 수 있게 되고, 따라서 onCreate에서 binding을 초기화해주는 것입니다.

## 2. Adapter의 종류
Adapter가 하는 역할은 세미나에서 말씀드렸던 것 처럼 데이터와 UI를 매핑시키는 것입니다.
사용자는 오직 데이터를 어떻게 UI로 만드는지 (getView 함수)만 코드로 짜고, 나머지 부분 (진짜 리스트로 그리는 부분)은 Adapter가 다 해줍니다. 즉, xml에 추가한 ListView와 데이터 간의 중개를 해 준다고 생각하면 될 것 같네요. 구현부는 저도 공부한 적은 없는데 굉장히 복잡할 거라 패스하셔도 됩니다. (평생 볼 일 없을지도...)

Adapter의 종류는 ListAdapter부터 상속 관계를 따라 위로 쭉 올라가서 Adapter.java에 도착한 후 어떤 것들이 얘를 상속하고 있나 살펴보면 알 수 있습니다. 다만 알아야 할 필요는 딱히 없을 것 같네요

<img width="951" alt="image" src="https://github.com/wafflestudio/seminar-2023/assets/88367636/f25c7df2-0e0a-41c3-8e26-5fdc06c32c07">

## 3. View의 Adapter
ppt에서는 View 중에서도 ListView였기 때문에 adapter라는 멤버 변수가 있는 것입니다. ListView.java의 코드를 보면 
```
    @Override
    public ListAdapter getAdapter() {
        return mAdapter;
    }
```
이런 부분이 있고, mAdapter는 ListView가 상속하는 AbsListView 추상 클래스가 갖는 프로퍼티네요. 
Adapter의 역할은 Adapter.java에서 Adapter라는 인터페이스가 어떤 함수를 제공하는지를 보면 어느 정도 알 수 있습니다.
[이곳](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/widget/Adapter.java?q=Adapter.java) 혹은 안드로이드 스튜디오에서 직접 볼 수 있는데요, 주석들을 잘 읽어보면(특히 getView) Adapter가 뭘 해 주는지 대략 알 수 있습니다. 구현부는 지금 이해할 수도 없고, 알 필요도 아직 없다고 생각합니다.(저도 모릅니다)

## 4. Adapter의 생성자의 list
생성자 부분은 정해진 게 전혀 없고, 제 맘대로 짠 것이라 상관 없습니다. List를 넘기셔도 됩니다. ppt 만들 때는 Array로 했는데 영상에는 list로 했던 것 같네요. 그냥 Adapter에 넘기고 싶은 데이터를 전달해 주는 것입니다. HashMap일 수도 있고 Set일 수도 있겠죠. **저희는 Adapter가 생성자로 전달받은 데이터를 View로 바꾸는 방법만 `getView` 함수에 구현해 주면 됩니다.**

## 5. 왜 생성자에 context를 넘겼는가?
아마 동적으로 UI를 만들려고 넘겼던 걸로 기억합니다. TypeA 데이터가 List<String>이었는데 리스트 원소별로 TextView를 하나씩 만들어서 LinearLayout에 추가하기 위해 context가 필요해서 전달했습니다. 그런데 context 전달 안 해줘도 `holder.itemView.context` 로 context를 가져올 수 있어서 없어도 될 것 같네요. 
정리 : Adapter의 생성자 부분은 100% 개인의 구현이다!

## 6. getView에서 각 파라미터의 의미

각 파라미터의 의미는 안스에서
<img width="788" alt="image" src="https://github.com/wafflestudio/seminar-2023/assets/88367636/7122937d-e35b-4cf1-816e-98a1eb1da1f4">
줄번호 오른쪽에 초록색 동그라미 I 랑 빨간 화살표 있는 걸 눌러 보시면 해당 함수가 뭘 override하고 있는지로 이동할 수 있는데요, 이동한 Adapter.java에 자세한 주석이 써 있으니 참고하시면 됩니다.

convertView의 의미는 녹화 영상이나 ppt에 보시면 RecyclerView 설명 직전에 ViewHolder 설명 부분에 나와있습니다.
parent는 말 그대로 해당 View의 parent입니다.


## Github Discussion
label은 maintainer만 붙일 수 있다고 해서, 질문만 올려주시면 저희가 붙이겠습니다.

**Reply by kkomul1** [reply]
자세한 설명 감사합니다!! 제가 상속 관계를 찾아보는 법이나 해당 부분의 내용을 찾아보는 방법에 아직 익숙치 않아 제가 해결할 수 있는 부분들이 여럿 있었는데 해결하지 못했던 것 같네요..ㅠㅠ 다음에는 더 찾아보고 질문하겠습니다!

**Reply by JuTaK97** [reply]
보통 줄번호 옆에 기호를 눌러서 상속 관계 위/아래로 이동할 수 있고 이건 단축키도 있습니다(윈도우는 단축키가 뭔지 모르겠네요). 정의와 사용처 왔다갔다 할 때는 ctrl+b도 많이 사용합니다.
java 파일 빨간줄 가득한거 보기 싫거나 더 전체 파일을 보고 싶으면 저는 https://cs.android.com/ 이 사이트를 이용하는 편입니다.

**Reply by kkomul1** [reply]
알려주셔서 감사합니다! 윈도우는 단축키가 Ctrl+U랑 Ctrl+Alt+B였네요~
그런데 한 가지 더 질문이 있어서 남깁니다..ㅎㅎ
RecyclerView의 Adapter에서 onCreateViewHolder 함수가 반환하는 값을 정할 때 ppt에서는 다음과 같은 방법을 사용했습니다.
TypeAViewHolder(TypeAItemViewBinding.inflate(LayoutInflater.from(parent.context)))
여기서 parent.context를 찾아보니 부모 RecyclerView의 context라는데, type_a_item_view.xml을 뷰로 만들기 위한 inflate를 할 때에는 type_a_item_view.xml에 관련된 Activity를 넘겨주어야 하는 거라고 생각했습니다. 그렇다면 parent.context가  type_a_item_view.xml에 관련된 Activity인건가요? Activity가 아닌 다른 context라면, type_a_item_view.xml과 관련된 녀석인가요?

**Reply by JuTaK97** [reply]
View는 내부에 context를 가지고 있습니다. View가 들어 있는 환경을 나타내는 것인데, 동적으로 생성할 때 직접 저희가 생성자로 넣어 주기도 하고 inflate 과정에서 생성될 때 프레임워크나 라이브러리 코드에서 넣어주기도 하겠죠.

액티비티에 붙어 있는 대부분 View들은 context가 그 액티비티입니다. 뷰를 만드는 라이브러리에 따라 Activity가 아니라 ContextThemeWrapper 라는 것일 때도 있는데, 아무튼 일반적인 수준에서 사용할 때는 큰 구분 없이 그냥 context로서 사용하면 됩니다.

parent가 이상한 다른 별에서 온 view가 아니라 현재 액티비티 상에 올바르게 존재하는 뷰이기 때문에, 그냥 parent의 context를 사용하면 MainActivity를 생성자로 받아서 사용하는 거랑 차이가 없습니다.
